### ISSUE TITLE
Parameter Decorator alternative for `@inject`

### ISSUE THREAD

**Issue by buzzysin** [uditalias/injex#15]
**Is your feature request related to a problem? Please describe.**

Wondering if it were possible in this library to do parameter dependency injection into the constructor instead of property dependency injection. Various articles seem to be against the latter though I am learning DI at the moment so I have no idea what the correct answer is.

**Describe the solution you'd like**

I **have not** tested this, but I ended up writing a small wrapper to implement this.

```ts
// src/lib/runtime/injex.ts
export const injectParam = (shadowedKey: string, dependencyNameOrType?: any) => {
  return function (target: any, _key: string | symbol, _index: number) {
    // _key is `undefined` in the constructor, I think, and _index doesn't
    // seem to be useful in guessing the value of `shadowedKey`
    inject(dependencyNameOrType)(target.prototype, shadowedKey);
  };
}

// Usage
import { Dependency } from "somewhere/in/my/project"

export class SomeService {
  constructor(
    // Here `dep` is required as I couldn't find a way to 
    // get the name of the paramter in the constructor using 
    // `reflect-metadata`
    @injectParam("dep", Dependency) private dep: Dependency) {}
  
  public someMethod() {
    const someResult = this.dep.someResource();
    // ...
  }
}
```

It's not great as two function parameters in the decorator go unused, but I'm sure someone more qualified can come up with a solution to that - maybe even detect the parameter names in the constructor.

**Describe alternatives you've considered**
[InversifyJS](https://github.com/inversify/InversifyJS) includes the desired functionality with the caveat that it requires explicit registration of dependencies which I am too lazy to consider (I am only building a small app at the moment, no need for overkill, methinks).

**Additional context**
What prompted this feature request:

![image](https://user-images.githubusercontent.com/41271986/109853117-78721300-7c4d-11eb-895d-76e2b01c46df.png)

**Comment by uditalias** [comment]
Thank you @buzzysin for a detailed intro.
I'll take a look at this, I believe we can add the `@injectParam()` decorator.

**Comment by buzzysin** [comment]
Great, I'm glad to help!

**Comment by uditalias** [comment]
@buzzysin
I finally added the `@injectParam()` decorator to Injex.
It's part of `v3` and currently released in `v3.0.0-alpha.6`.

**Note**
Since the dependencies resolved after class creation in Injex, you can't use this feature in the constructor. You can use the `@init()` decorator to set a method as an initializer.

You can read more about `@injectParam()` and how to use it here:
https://www.injex.dev/docs/api/core/decorators/inject-param
